--- a/configs/nanopi_neo_air_defconfig
+++ b/configs/nanopi_neo_air_defconfig
@@ -9,3 +9,11 @@ CONFIG_DRAM_CLK=408
 CONFIG_CONSOLE_MUX=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_OHCI_HCD=y
+CONFIG_BOOTDELAY=0
+CONFIG_SUPPORT_EMMC_BOOT=y
+CONFIG_MMC_SUNXI_SLOT_EXTRA=2
+
+CONFIG_SPI=y
+CONFIG_SPL_SPI_SUNXI=y
+CONFIG_CMD_SF=y
+CONFIG_SPI_FLASH_WINBOND=y
--- a/arch/arm/dts/sun8i-h3-nanopi-neo-air.dts
+++ b/arch/arm/dts/sun8i-h3-nanopi-neo-air.dts
@@ -141,3 +141,16 @@
 	/* USB VBUS is always on */
 	status = "okay";
 };
+
+&spi0 {
+	status = "okay";
+	spi-max-frequency = <12000000>;
+
+	flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "winbond,w25q128", "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <12000000>;
+	};
+};
--- a/include/configs/sunxi-common.h
+++ b/include/configs/sunxi-common.h
@@ -228,6 +228,26 @@
 #define BOOT_TARGET_DEVICES_USB(func)
 #endif
 
+#ifdef CONFIG_CMD_SF
+#define BOOTENV_DEV_NAME_SF(devtypeu, devtypel, instance) \
+	#devtypel #instance " "
+#define BOOTENV_DEV_SF(devtypeu, devtypel, instance) \
+	"bootcmd_" #devtypel #instance "=" \
+		"setenv busnum " #instance "; " \
+		"run " #devtypel "_boot\0" \
+	#devtypel "_boot=if " #devtypel " probe ${busnum}; then " \
+		"devtype=" #devtypel "; " \
+		"run scan_sf_for_scripts; " \
+	"fi\0" \
+	"scan_sf_for_scripts=" \
+		#devtypel " read ${scriptaddr} ${script_offset_f} ${script_size_f}; " \
+		"source ${scriptaddr}; " \
+		"echo SCRIPT FAILED: continuing...\0"
+#define BOOT_TARGET_DEVICES_SF(func) func(SF, sf, 0)
+#else
+#define BOOT_TARGET_DEVICES_SF(func)
+#endif
+
 #ifdef CONFIG_CMD_PXE
 #define BOOT_TARGET_DEVICES_PXE(func) func(PXE, pxe, na)
 #else
@@ -252,6 +272,7 @@
 
 #define BOOT_TARGET_DEVICES(func) \
 	func(FEL, fel, na) \
+	BOOT_TARGET_DEVICES_SF(func) \
 	BOOT_TARGET_DEVICES_MMC(func) \
 	BOOT_TARGET_DEVICES_SCSI(func) \
 	BOOT_TARGET_DEVICES_USB(func) \
@@ -325,10 +346,12 @@
 	DFU_ALT_INFO_RAM \
 	"fdtfile=" FDTFILE "\0" \
 	"console=ttyS0,115200\0" \
+	"script_offset_f=0xFA000\0" \
+	"script_size_f=0x2000\0" \
 	"uuid_gpt_esp=" UUID_GPT_ESP "\0" \
 	"uuid_gpt_system=" UUID_GPT_SYSTEM "\0" \
 	"partitions=" PARTS_DEFAULT "\0" \
 	BOOTCMD_SUNXI_COMPAT \
 	BOOTENV
 
-#endif /* _SUNXI_COMMON_CONFIG_H */
+#endif /* _SUNXI_COMMON_CONFIG_H */
\ No newline at end of file
