--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -4824,7 +4824,11 @@ void kick_defer_list_purge(struct softne
 		backlog_unlock_irq_restore(sd, &flags);
 
 	} else if (!cmpxchg(&sd->defer_ipi_scheduled, 0, 1)) {
+#ifndef CONFIG_PREEMPT_RT
 		smp_call_function_single_async(cpu, &sd->defer_csd);
+#else
+		schedule_work_on(cpu, &sd->defer_work);
+#endif
 	}
 }
 
--- a/net/core/page_pool.c
+++ b/net/core/page_pool.c
@@ -897,7 +897,11 @@ static void page_pool_release_retry(stru
 		if (cpu == raw_smp_processor_id())
 			raise_softirq_irqoff(NET_RX_SOFTIRQ);
 		else if (!cmpxchg(&sd->defer_ipi_scheduled, 0, 1))
+#ifndef CONFIG_PREEMPT_RT
 			smp_call_function_single_async(cpu, &sd->defer_csd);
+#else
+			schedule_work_on(cpu, &sd->defer_work);
+#endif
 	}
 	local_irq_restore(flags);
 
